filesystem/android/src/main/java/com/capacitorjs/plugins/filesystem/FilesystemPlugin.java

```diff
@@ -0,0 +1,494 @@
+package com.capacitorjs.plugins.filesystem;
+
+import android.Manifest;
+import android.content.pm.PackageManager;
+import android.media.MediaScannerConnection;
+import android.net.Uri;
+import com.capacitorjs.plugins.filesystem.exceptions.CopyFailedException;
+import com.capacitorjs.plugins.filesystem.exceptions.DirectoryExistsException;
+import com.capacitorjs.plugins.filesystem.exceptions.DirectoryNotFoundException;
+import com.getcapacitor.JSArray;
+import com.getcapacitor.JSObject;
+import com.getcapacitor.Logger;
+import com.getcapacitor.Plugin;
+import com.getcapacitor.PluginCall;
+import com.getcapacitor.PluginMethod;
+import com.getcapacitor.PluginRequestCodes;
+import com.getcapacitor.annotation.CapacitorPlugin;
+import com.getcapacitor.annotation.Permission;
+import java.io.*;
+import java.nio.charset.Charset;
+import org.json.JSONException;
+
+@CapacitorPlugin(
+    name = "Filesystem",
+    requestCodes = {
+        PluginRequestCodes.FILESYSTEM_REQUEST_WRITE_FILE_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_WRITE_FOLDER_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_READ_FILE_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_READ_FOLDER_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_DELETE_FILE_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_DELETE_FOLDER_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_URI_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_STAT_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_RENAME_PERMISSIONS,
+        PluginRequestCodes.FILESYSTEM_REQUEST_COPY_PERMISSIONS
+    },
+    permissions = {
+        @Permission(permission = Manifest.permission.READ_EXTERNAL_STORAGE, alias = "publicStorage"),
+        @Permission(permission = Manifest.permission.WRITE_EXTERNAL_STORAGE, alias = "publicStorage")
```

Comment:
The permission is only needed if you try to access the public storage (shared), so I called it publicStorage instead of going for just "storage" as I think that might cause confusion and make users request it every time despite they don't need it for non public storage.
But if you have a better name, I'm open for suggestions

Reply:
My initial thought was `externalStorage`, but if `publicStorage` is more clear (and more cross-platform), then it's good 👍 I was mostly just curious

Reply:
there are some external storage locations that don't need permission. The code had a `isPublicDirectory` function that checks if it's one of the locations that need the permission, that's why I also used public on the name.